hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/lock/WrappedStripedLock.java

```diff
@@ -0,0 +1,177 @@
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import org.apache.hadoop.hdds.utils.SimpleStriped;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Locking class that wraps Striped locking with timeout and logging. It also provides ordering while multiple locks.
+ */
+public class WrappedStripedLock {
+  private static final Logger LOG = LoggerFactory.getLogger(WrappedStripedLock.class);
+  private final Striped<ReadWriteLock> fileStripedLock;
+  private final long lockTimeout;
+
+  public WrappedStripedLock(int stripLockSize, long timeout, boolean fair) {
+    fileStripedLock = SimpleStriped.readWriteLock(stripLockSize, fair);
+    lockTimeout = timeout;
+  }
+
+  /**
+   * lock the list of keys in order.
+   * Sample code for lock and unlock handling:
+   * <code>
+   *   try {
+   *     if (!wrappedStripedLock.lock(keyList, locks)) {
+   *       // timeout occurred, release lock if any in reverse order
+   *       Collections.reverse(locks);
+   *       locks.forEach(Lock::unlock);
+   *     }
+   *     // perform business logic
+   *   } finally {
+   *     // to be released in reverse order
+   *     Collections.reverse(locks);
+   *     locks.forEach(Lock::unlock);
+   *   }
+   * </code>
+   * 
+   * @param keyList key list which needs to be locked
+   * @param locks successful lock object returned which will be used to release lock
+   * @return boolean true if success, else false
+   * @throws InterruptedException exception on interrupt
+   */
+  public boolean writeLock(List<String> keyList, List<Lock> locks) throws InterruptedException {
+    try {
+      Iterable<ReadWriteLock> readWriteLocks = fileStripedLock.bulkGet(keyList);
+      for (ReadWriteLock rwLock : readWriteLocks) {
+        Lock lockObj = rwLock.writeLock();
+        boolean b = lockObj.tryLock(lockTimeout, TimeUnit.MILLISECONDS);
+        if (!b) {
+          LOG.error("Write lock for keys are failed for the instance {} after wait of {}ms, read lock info: {}", this,
+              lockTimeout, rwLock.readLock());
+          return false;
+        }
+        locks.add(lockObj);
+      }
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      LOG.error("Write lock for keys are interrupted for the instance {}", this);
+      throw e;
+    }
+    return true;
+  }
+
+  /**
+   * lock single key.
+   * @param key object for which lock to be taken
+   * @return lock object to be used to release lock, null if unable to take lock due to timeout
+   * @throws InterruptedException exception on interrupt
+   */
+  public Lock writeLock(String key) throws InterruptedException {
+    LOG.debug("Key {} is locked for instance {} {}", key, this, fileStripedLock.get(key));
+    try {
+      Lock lockObj = fileStripedLock.get(key).writeLock();
+      boolean b = lockObj.tryLock(lockTimeout, TimeUnit.MILLISECONDS);
```

Comment:
Do not mixing getLock and acquireLock together.  It makes the code complicated and error prone.

Reply:
it first get() lock object based on name belonging to the bucket, which internally it generate hash and return.
Next is, tryLock() to lock
I am not getting the problem.

Reply:
Then, how would you release the lock?  get it again and then release?  Why getting the same lock twice?